heat/utils/data/datatools.py

```diff
@@ -0,0 +1,211 @@
+            self._first_iter = False
+        else:
+            # shuffle after the first epoch but before the iterator is generated
+            self.shuffle()
+        return self.DataLoader.__iter__()
+
+    def __len__(self) -> int:
+        return len(self.DataLoader)
+
+    def shuffle(self):
+        """
+        Shuffle the data. This is a direct call to the shuffle function of the dataset.
+        This function is blocking.
+        """
+        self.dataset.shuffle()
+
+
+class Dataset(torch_data.Dataset):
+    """
+    An abstract class representing a given dataset. This inherits from torch.utils.data.Dataset.
+
+    This class is a general example for what should be done to create a Dataset. When creating a dataset all of the
+    standard attributes should be set, the ``__getitem__``, ``__len__``, and ``shuffle`` functions must be defined.
+
+        - ``__getitem__`` : how an item is given to the network
+        - ``__len__`` : the number of data elements to be given to the network in total
+        - ``shuffle`` : how the data should be shuffled between the processes. The function shown below is for a dataset
+            composed of only data and without targets
+
+    As the amount of data across processes can be non-uniform, the dataset class will slice off the remaining elements
+    on whichever processes have more data than the others. This should only be 1 element.
+    The shuffle function will shuffle all of the data on the process.
+
+    It is recommended that for DNDarrays, the split is either 0 or None
+
+    Parameters
+    ----------
+    array : DNDarray
+        DNDarray for which to great the dataset
+    transform : Callable
+        transformation to call before a data item is returned
+
+    Attributes
+    ----------
+    These are the required attributes. Optional attributed are whatever are required for the Dataset
+    (see :class:`heat.utils.data.mnist.py`)
+    htdata : DNDarray
+        full data
+    _cut_slice : slice
+        slice to cut off the last element to get a uniform amount of data on each process
+    comm : MPICommunicator
+        communication object used to send the data between processes
+    lcl_half : int
+        half of the number of data elements on the process
+    data : torch.Tensor
+        the local data to be used in training
+    transform : Callable
+        transform to be called during the getitem function
+
+    TODO: type annotation for array
+    """
+
+    def __init__(self, array, transform: Callable = None):
+        self.htdata = array
+        self.comm = array.comm
+        # create a slice to create a uniform amount of data on each process
+        min_data_split = array.gshape[array.split] // array.comm.size
+        self.lcl_half = min_data_split // 2
+        arb_slice = [slice(None)] * array.ndim
+        arb_slice[array.split] = slice(min_data_split)
+        self._cut_slice = tuple(arb_slice)
+        self.data = array._DNDarray__array[self._cut_slice]
```

Comment:
not necessarily. This is not a complex operation and it only happens on each process. 

What is happening here is the last element is cut off from `self.htdata` if the process has more elements than the others. After this is done, `self.data` is a local torch tensor which is the same size on every process. this ensures that the batches used on every process are the same size. A torch tensor is used here because it is what is needed by the torch DataLoader which is used to feed data into the network and to automatically determine batch sizes